File: ledger_bot.js
const axios = require('axios');
const CryptoJS = require('crypto-js');
const { GoogleSpreadsheet } = require('google-spreadsheet');
const { JWT } = require('google-auth-library');
const fs = require('fs');
require('dotenv').config();

// === 설정 ===
const CONFIG = {
  groupSeq: 'gcmsAmaranth39483',
  callerName: 'API_gcmsAmaranth39483',
  accessToken: 'roifHrayJttms27ufGiqVa8grv6Sk0',
  hashKey: '88761859188784596178355689527478836553536918',
  proxyUrl: '/apiproxy/api11A30',
  amaranthUrl: 'https://portal.rapportlabs.kr',
  coCd: '1000',
  sheetId: '1jcO4dHExbdwT6sZejj2Z22pycvZ6dRsyqPZ62zgUk-Y',
  sheetTabName: '계정별원장_RAW'
};

// 수정된 계정 리스트 (입력불가 제외)
const SGA_ACCOUNTS = [
  '8000000', '8010000', '8020000', '8020001', '8030000', '8040000', '8050000',
  '8060000', '8070000', '8080000', '8090000', 
  '8110000', '8110001', '8110002', '8110003', '8110004', '8110005', '8110006',
  '8120000', '8130000', '8140000', '8150000', '8160000', '8170000', '8180000', 
  '8190000', '8200000', '8210000', '8220000', '8240000', '8250000', '8260000', 
  '8270000', '8280000', '8290000', '8300000', '8300001', '8300002',
  '8310001', '8310002', '8310003', '8310004', '8310005', '8310006', '8310007', '8310008',
  '8330000', '8330001', 
  '8340000', '8340001', '8340002', '8340003', '8340004', '8340005', '8340006', '8340007', '8340008',
  '8350000', '8360000', '8380000', '8390000', 
  '8410000', '8420000', '8430000', '8440000', '8450000', '8460000', '8470000',
  '8480000', '8480001', '8480002', '8480003', 
  '8490000', '8500000', '8510000'
];

// Axios 인스턴스 생성 (타임아웃 설정)
const apiClient = axios.create({
  baseURL: CONFIG.amaranthUrl,
  timeout: 30000, // 30초 타임아웃
  headers: {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
  }
});

// 유틸리티 함수들
function generateTransactionId(length = 30) {
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  let result = '';
  for (let i = 0; i < length; i++) result += chars.charAt(Math.floor(Math.random() * chars.length));
  return result;
}

function generateWehagoSign(hashKey, value) {
  const hmac = CryptoJS.HmacSHA256(value, hashKey);
  return CryptoJS.enc.Base64.stringify(hmac);
}

function getTodayString() {
  const today = new Date();
  const year = today.getFullYear();
  const month = String(today.getMonth() + 1).padStart(2, '0');
  const day = String(today.getDate()).padStart(2, '0');
  return year + month + day;
}

function formatDateStr(dt) {
  if (!dt || dt.length !== 8) return dt || '';
  return dt.substring(0, 4) + '-' + dt.substring(4, 6) + '-' + dt.substring(6, 8);
}

function convertDrcrFg(fg) {
  if (fg === '1') return '차변';
  if (fg === '2') return '대변';
  return fg;
}

// 날짜 기간 분할 (1개월 단위)
function getMonthlyRanges(startStr, endStr) {
  const ranges = [];
  const parseDate = (str) => new Date(str.substring(0, 4), str.substring(4, 6) - 1, str.substring(6, 8));
  const stringifyDate = (date) => {
    const y = date.getFullYear();
    const m = String(date.getMonth() + 1).padStart(2, '0');
    const d = String(date.getDate()).padStart(2, '0');
    return `${y}${m}${d}`;
  };

  let current = parseDate(startStr);
  const end = parseDate(endStr);

  while (current <= end) {
    const rangeStart = new Date(current);
    const nextMonth = new Date(current);
    nextMonth.setMonth(current.getMonth() + 1);
    nextMonth.setDate(1);
    
    let rangeEnd = new Date(nextMonth);
    rangeEnd.setDate(rangeEnd.getDate() - 1);
    
    if (rangeEnd > end) rangeEnd = new Date(end);

    ranges.push({ from: stringifyDate(rangeStart), to: stringifyDate(rangeEnd) });
    current = new Date(nextMonth);
  }
  return ranges;
}

// API 호출
async function callAccountLedgerApi(params) {
  const transactionId = generateTransactionId();
  const timestamp = Math.floor(Date.now() / 1000).toString();
  const url = CONFIG.proxyUrl;
  
  const wehagoSign = generateWehagoSign(
    CONFIG.hashKey,
    CONFIG.accessToken + transactionId + timestamp + url
  );
  
  const requestBody = {
    header: { groupSeq: CONFIG.groupSeq, empSeq: 1, tId: '', pId: '' },
    coCd: params.coCd,
    divCds: params.divCds,
    fillDtFrom: params.fillDtFrom,
    fillDtTo: params.fillDtTo,
    prtFg: params.prtFg,
    acctCd: params.acctCd,
    zeroDisp: params.zeroDisp,
    viewPage: params.viewPage,
    viewCount: params.viewCount
  };
  
  try {
    const response = await apiClient.post(CONFIG.proxyUrl, requestBody, {
      headers: {
        'callerName': CONFIG.callerName,
        'Authorization': 'Bearer ' + CONFIG.accessToken,
        'transaction-id': transactionId,
        'timestamp': timestamp,
        'groupSeq': CONFIG.groupSeq,
        'wehago-sign': wehagoSign,
        'Content-Type': 'application/json'
      }
    });
    return response.data;
  } catch (error) {
    if (error.code === 'ECONNABORTED') {
      console.error(`❌ [Timeout] ${params.acctCd}: 요청 시간 초과`);
    } else {
      console.error(`❌ [Error] ${params.acctCd}: ${error.message}`);
    }
    return { resultCode: -1, resultMsg: error.message };
  }
}

// 페이지 순회 (수정된 로직)
async function fetchAllPagesForAccount(params) {
  const allData = [];
  let currentPage = 1;
  
  while (true) {
    params.viewPage = currentPage;
    
    const result = await callAccountLedgerApi(params);
    
    if (result.resultCode !== 0) break; // 에러 시 중단
    
    const resultData = result.resultData;
    if (!resultData || !resultData.datas || resultData.datas.length === 0) {
      break; // 데이터 없음
    }
    
    allData.push(...resultData.datas);
    
    // 총 페이지 수 도달 시 종료
    if (currentPage >= (resultData.totalPage || 1)) {
      break;
    }
    
    currentPage++;
    await new Promise(resolve => setTimeout(resolve, 100)); // 0.1초 대기
  }
  
  return allData;
}

// 구글 시트 업로드
async function uploadToGoogleSheet(data) {
  if (data.length === 0) {
    console.log('⚠️ 업로드할 데이터가 없습니다.');
    return;
  }

  console.log('📊 구글 시트 연결 중...');
  try {
    const serviceAccountAuth = new JWT({
      email: require('./service_account.json').client_email,
      key: require('./service_account.json').private_key,
      scopes: ['https://www.googleapis.com/auth/spreadsheets'],
    });

    const doc = new GoogleSpreadsheet(CONFIG.sheetId, serviceAccountAuth);
    await doc.loadInfo();
    
    let sheet = doc.sheetsByTitle[CONFIG.sheetTabName];
    if (!sheet) {
      sheet = await doc.addSheet({ title: CONFIG.sheetTabName, headerValues: [] });
    }
    
    await sheet.clear();
    
    const headers = [
      '회사코드', '사업장코드', '계정과목', '차대구분', '승인일', '승인번호',
      '적요', '거래처코드', '거래처명', '사업자번호', '차변', '대변', '잔액',
      '작성일', '작성순번', '화면순번', '라인순번',
      '사용부서코드', '사용부서명', '프로젝트코드', '프로젝트명', '사용사원코드', '사용사원명'
    ];
    
    await sheet.setHeaderRow(headers);
    
    const rows = data.map(item => ({
      '회사코드': item.coCd || '', '사업장코드': item.divCd || '', '계정과목': item.acctCd || '',
      '차대구분': convertDrcrFg(item.drcrFg), '승인일': formatDateStr(item.fillDt), '승인번호': item.fillNb || 0,
      '적요': item.rmkDc || '', '거래처코드': item.trCd || '', '거래처명': item.trNm || '', '사업자번호': item.regNb || '',
      '차변': item.drAm || 0, '대변': item.crAm || 0, '잔액': item.restAm || 0,
      '작성일': formatDateStr(item.isuDt), '작성순번': item.isuSq || 0, '화면순번': item.dispSq || 0, '라인순번': item.lnSq || 0,
      '사용부서코드': item.ctDeptCd || '', '사용부서명': item.ctDeptNm || '',
      '프로젝트코드': item.pjtCd || '', '프로젝트명': item.pjtNm || '',
      '사용사원코드': item.ctEmpCd || '', '사용사원명': item.ctEmpNm || ''
    }));
    
    rows.sort((a, b) => (a['승인일'] !== b['승인일'] ? (a['승인일'] < b['승인일'] ? -1 : 1) : a['승인번호'] - b['승인번호']));

    const CHUNK_SIZE = 2000;
    for (let i = 0; i < rows.length; i += CHUNK_SIZE) {
      const chunk = rows.slice(i, i + CHUNK_SIZE);
      await sheet.addRows(chunk);
      console.log(`  - ${Math.min(i + CHUNK_SIZE, rows.length)} / ${rows.length} 건 업로드 완료`);
    }
    
    await sheet.loadCells('Y1:Y2');
    const headerCell = sheet.getCell(0, 24);
    const valueCell = sheet.getCell(1, 24);
    headerCell.value = '업데이트'; headerCell.textFormat = { bold: true };
    const now = new Date();
    valueCell.value = now.toISOString().replace('T', ' ').substring(0, 19);
    await sheet.saveUpdatedCells();
    
    console.log('✅ 구글 시트 업로드 완료!');
  } catch (error) {
    console.error('❌ 구글 시트 업로드 오류:', error);
  }
}

// 메인 실행
async function main() {
  const startDate = '20250101';
  const endDate = getTodayString();
  const dateRanges = getMonthlyRanges(startDate, endDate);
  
  console.log(`=== 판관비 계정별원장 조회 시작 (${startDate} ~ ${endDate}) ===`);
  console.log(`📅 총 ${dateRanges.length}개 구간 (월 단위)`);

  const baseParams = {
    coCd: CONFIG.coCd,
    divCds: '1000|',
    prtFg: '2',
    zeroDisp: '0',
    viewCount: 50
  };
  
  const allData = [];
  
  for (let i = 0; i < SGA_ACCOUNTS.length; i++) {
    const acctCd = SGA_ACCOUNTS[i];
    let accountTotal = 0;
    
    for (const range of dateRanges) {
      try {
        const params = { ...baseParams, acctCd: acctCd, fillDtFrom: range.from, fillDtTo: range.to };
        const periodData = await fetchAllPagesForAccount(params);
        if (periodData.length > 0) {
          allData.push(...periodData);
          accountTotal += periodData.length;
        }
      } catch (error) {
        console.error(`❌ ${acctCd} (${range.from}): ${error.message}`);
      }
    }
    
    if (accountTotal > 0) {
      console.log(`[${i+1}/${SGA_ACCOUNTS.length}] ${acctCd}: ✅ ${accountTotal}건`);
    } else {
      console.log(`[${i+1}/${SGA_ACCOUNTS.length}] ${acctCd}: 데이터 없음`);
    }
    
    await new Promise(resolve => setTimeout(resolve, 50)); 
  }
  
  console.log(`\n=== 조회 완료 (총 ${allData.length}건) ===`);
  await uploadToGoogleSheet(allData);
}

main();

File: ledger_bot.py
import os
import time
import hmac
import hashlib
import base64
import random
import string
import json
import requests
import pandas as pd
import gspread
from datetime import datetime
from google.oauth2.service_account import Credentials
from logger import logger  # 기존 로거 사용

# === 설정값 (보안을 위해 .env 관리를 권장합니다) ===
CONFIG = {
    'groupSeq': 'gcmsAmaranth39483',
    'callerName': 'API_gcmsAmaranth39483',
    'accessToken': 'roifHrayJttms27ufGiqVa8grv6Sk0',
    'hashKey': '88761859188784596178355689527478836553536918',
    'proxyUrl': '/apiproxy/api11A30',
    'amaranthUrl': 'https://portal.rapportlabs.kr',
    'coCd': '1000',
    'sheetId': '1jcO4dHExbdwT6sZejj2Z22pycvZ6dRsyqPZ62zgUk-Y',
    'sheetTabName': '계정별원장_RAW'
}

# 판관비 계정과목 목록
SGA_ACCOUNTS = [
    '8000000', '8010000', '8020000', '8020001', '8030000', '8040000', '8050000',
    '8060000', '8070000', '8080000', '8090000', '8100000', '8110000', '8110001',
    '8110002', '8110003', '8110004', '8110005', '8110006', '8120000', '8130000',
    '8140000', '8150000', '8160000', '8170000', '8180000', '8190000', '8200000',
    '8210000', '8220000', '8230000', '8240000', '8250000', '8260000', '8270000',
    '8280000', '8290000', '8300000', '8300001', '8300002', '8310000', '8310001',
    '8310002', '8310003', '8310004', '8310005', '8310006', '8310007', '8310008',
    '8320000', '8330000', '8330001', '8340000', '8340001', '8340002', '8340003',
    '8340004', '8340005', '8340006', '8340007', '8340008', '8350000', '8360000',
    '8370000', '8380000', '8390000', '8400000', '8410000', '8420000', '8430000',
    '8440000', '8450000', '8460000', '8470000', '8480000', '8480001', '8480002',
    '8480003', '8490000', '8500000', '8510000'
]

def generate_transaction_id(length=30):
    """30자리 랜덤 문자열 생성"""
    chars = string.ascii_lowercase + string.digits
    return ''.join(random.choice(chars) for _ in range(length))

def generate_wehago_sign(hash_key, value):
    """Wehago 서명 생성 (HMAC-SHA256 -> Base64)"""
    # Python의 hmac은 bytes 타입을 요구합니다.
    key_bytes = hash_key.encode('utf-8')
    value_bytes = value.encode('utf-8')
    
    signature = hmac.new(key_bytes, value_bytes, hashlib.sha256).digest()
    return base64.b64encode(signature).decode('utf-8')

def get_today_string():
    """오늘 날짜 yyyymmdd"""
    return datetime.now().strftime('%Y%m%d')

def format_date_str(dt_str):
    """yyyymmdd -> yyyy-mm-dd 변환"""
    if not dt_str or len(str(dt_str)) != 8:
        return dt_str
    s = str(dt_str)
    return f"{s[:4]}-{s[4:6]}-{s[6:]}"

def convert_drcr_fg(fg):
    """차대구분 코드 변환"""
    if fg == '1': return '차변'
    if fg == '2': return '대변'
    return fg

def call_account_ledger_api(params):
    """API 호출 함수"""
    transaction_id = generate_transaction_id()
    timestamp = str(int(time.time()))
    url_path = CONFIG['proxyUrl']
    
    # 서명 생성 값: accessToken + transactionId + timestamp + url
    sign_value = CONFIG['accessToken'] + transaction_id + timestamp + url_path
    wehago_sign = generate_wehago_sign(CONFIG['hashKey'], sign_value)
    
    api_url = CONFIG['amaranthUrl'] + url_path
    
    request_body = {
        "header": {
            "groupSeq": CONFIG['groupSeq'],
            "empSeq": 1,
            "tId": "",
            "pId": ""
        },
        "coCd": params.get('coCd'),
        "divCds": params.get('divCds'),
        "fillDtFrom": params.get('fillDtFrom'),
        "fillDtTo": params.get('fillDtTo'),
        "prtFg": params.get('prtFg'),
        "acctCd": params.get('acctCd'),
        "zeroDisp": params.get('zeroDisp'),
        "viewPage": params.get('viewPage'),
        "viewCount": params.get('viewCount')
    }
    
    headers = {
        'callerName': CONFIG['callerName'],
        'Authorization': 'Bearer ' + CONFIG['accessToken'],
        'transaction-id': transaction_id,
        'timestamp': timestamp,
        'groupSeq': CONFIG['groupSeq'],
        'wehago-sign': wehago_sign,
        'Content-Type': 'application/json'
    }
    
    try:
        # timeout=30 추가 (30초 대기)
        response = requests.post(api_url, headers=headers, json=request_body, verify=True, timeout=30)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.Timeout:
        logger.error(f"API Request Timeout (30s)")
        return {"resultCode": -1, "resultMsg": "Timeout"}
    except requests.exceptions.RequestException as e:
        logger.error(f"API Request Failed: {e}")
        return {"resultCode": -1, "resultMsg": str(e)}

def fetch_all_pages_for_account(base_params):
    """특정 계정의 전체 페이지 데이터 수집"""
    all_data = []
    current_page = 1
    total_page = 1
    
    while current_page <= total_page:
        base_params['viewPage'] = current_page
        result = call_account_ledger_api(base_params)
        
        if result.get('resultCode') != 0:
            logger.error(f"API Error for {base_params['acctCd']}: {result.get('resultMsg')}")
            return []
            
        result_data = result.get('resultData', {})
        if not result_data:
            return all_data
            
        total_page = result_data.get('totalPage', 1)
        datas = result_data.get('datas', [])
        
        if datas:
            all_data.extend(datas)
            
        current_page += 1
        # API 과부하 방지를 위해 짧은 대기
        if current_page <= total_page:
            time.sleep(0.1)
            
    return all_data

def upload_to_google_sheet(data_list):
    """데이터프레임을 구글 시트에 업로드"""
    if not data_list:
        logger.warning("업로드할 데이터가 없습니다.")
        return

    try:
        logger.info("📊 구글 시트 연결 중...")
        
        # 서비스 계정 인증
        # service_account.json 파일이 같은 경로에 있어야 합니다.
        gc = gspread.service_account(filename='service_account.json')
        sh = gc.open_by_key(CONFIG['sheetId'])
        
        try:
            worksheet = sh.worksheet(CONFIG['sheetTabName'])
            worksheet.clear()
        except gspread.WorksheetNotFound:
            worksheet = sh.add_worksheet(title=CONFIG['sheetTabName'], rows=1000, cols=26)
            
        # Pandas로 데이터 가공
        df = pd.DataFrame(data_list)
        
        # 필요한 컬럼 매핑 및 순서 정렬
        # GAS 코드의 headers 순서와 맞춤
        columns_map = {
            'coCd': '회사코드', 'divCd': '사업장코드', 'acctCd': '계정과목', 'drcrFg': '차대구분',
            'fillDt': '승인일', 'fillNb': '승인번호', 'rmkDc': '적요', 'trCd': '거래처코드',
            'trNm': '거래처명', 'regNb': '사업자번호', 'drAm': '차변', 'crAm': '대변',
            'restAm': '잔액', 'isuDt': '작성일', 'isuSq': '작성순번', 'dispSq': '화면순번',
            'lnSq': '라인순번', 'ctDeptCd': '사용부서코드', 'ctDeptNm': '사용부서명',
            'pjtCd': '프로젝트코드', 'pjtNm': '프로젝트명', 'ctEmpCd': '사용사원코드',
            'ctEmpNm': '사용사원명'
        }
        
        # 존재하는 컬럼만 선택하여 이름 변경
        target_cols = [col for col in columns_map.keys() if col in df.columns]
        df = df[target_cols].rename(columns=columns_map)
        
        # 데이터 포맷팅
        if '차대구분' in df.columns:
            df['차대구분'] = df['차대구분'].apply(convert_drcr_fg)
        if '승인일' in df.columns:
            df['승인일'] = df['승인일'].apply(format_date_str)
        if '작성일' in df.columns:
            df['작성일'] = df['작성일'].apply(format_date_str)
            
        # 숫자형 변환 (NaN -> 0)
        numeric_cols = ['승인번호', '차변', '대변', '잔액', '작성순번', '화면순번', '라인순번']
        for col in numeric_cols:
            if col in df.columns:
                df[col] = pd.to_numeric(df[col], errors='coerce').fillna(0)

        # 정렬 (승인일, 승인번호)
        if '승인일' in df.columns and '승인번호' in df.columns:
            df = df.sort_values(by=['승인일', '승인번호'])

        # 업데이트 시간 컬럼 추가 (마지막 컬럼 뒤에)
        update_time = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        # 헤더 + 데이터 준비
        headers = df.columns.tolist()
        values = df.astype(str).values.tolist() # gspread 호환을 위해 string 변환
        
        # 시트 업데이트
        worksheet.update(range_name='A1', values=[headers] + values)
        
        # 업데이트 시간 별도 표기 (헤더 옆)
        worksheet.update_cell(1, len(headers) + 2, "업데이트")
        worksheet.update_cell(2, len(headers) + 2, update_time)
        
        logger.info(f"✅ 시트 업로드 완료: {len(values)}건")
        
    except Exception as e:
        logger.error(f"❌ 시트 업로드 실패: {e}")

def run_ledger_bot():
    """메인 실행 함수"""
    
    base_params = {
        'coCd': CONFIG['coCd'],
        'divCds': '1000|',
        'fillDtFrom': '20250101',
        'fillDtTo': get_today_string(),
        'prtFg': '2',
        'zeroDisp': '0',
        'viewPage': 1,
        'viewCount': 100
    }
    
    logger.info(f"=== 판관비 계정별원장 조회 시작 ({base_params['fillDtFrom']} ~ {base_params['fillDtTo']}) ===")
    
    all_data = []
    success_count = 0
    empty_count = 0
    
    for i, acct_cd in enumerate(SGA_ACCOUNTS):
        try:
            logger.info(f"[{i+1}/{len(SGA_ACCOUNTS)}] {acct_cd} 조회 시작...") # 진행 상황 로그 추가
            params = base_params.copy()
            params['acctCd'] = acct_cd
            
            account_data = fetch_all_pages_for_account(params)
            
            if account_data:
                all_data.extend(account_data)
                logger.info(f"[{i+1}/{len(SGA_ACCOUNTS)}] {acct_cd}: ✅ {len(account_data)}건")
                success_count += 1
            else:
                # logger.debug(f"[{i+1}/{len(SGA_ACCOUNTS)}] {acct_cd}: 데이터 없음")
                logger.info(f"[{i+1}/{len(SGA_ACCOUNTS)}] {acct_cd}: 데이터 없음") # 빈 것도 로그 출력
                empty_count += 1
                
            time.sleep(0.1) # 루프 간 짧은 대기
            
        except Exception as e:
            logger.error(f"[{i+1}/{len(SGA_ACCOUNTS)}] {acct_cd}: ❌ {e}")

    logger.info(f"\n=== 조회 완료 ===")
    logger.info(f"총 데이터: {len(all_data)}건 (유효 계정: {success_count}개)")
    
    if all_data:
        upload_to_google_sheet(all_data)
    else:
        logger.warning("⚠️ 조회된 데이터가 없어 시트를 업데이트하지 않았습니다.")

if __name__ == "__main__":
    run_ledger_bot()

